### ISSUE TITLE
Pax multiprocessing crash ProcessBatchQueue

### ISSUE THREAD

**Issue by pdeperio** [XENON1T/pax#463]
_From @lucrlom on November 24, 2016 17:5_

I had this problem at least in two runs 4696 and 4700 processing with `massive-cax --once --run 4xxx` with pax_v6.1.0 environment:
```
`cax_v4.10.1 - 2016-11-24 04:15:23,038 [CRITICAL] Exception caught from task ProcessBatchQueue
Traceback (most recent call last):
  File "/project/lgrandi/anaconda3/envs/pax_v6.1.0/lib/python3.4/site-packages/cax-4.10.1-py3.4.egg/cax/main.py", line 120, in main
    task.go(args.run)
  File "/project/lgrandi/anaconda3/envs/pax_v6.1.0/lib/python3.4/site-packages/cax-4.10.1-py3.4.egg/cax/task.py", line 65, in go
    self.each_run()
  File "/project/lgrandi/anaconda3/envs/pax_v6.1.0/lib/python3.4/site-packages/cax-4.10.1-py3.4.egg/cax/tasks/process.py", line 211, in each_run
    ncpus)
  File "/project/lgrandi/anaconda3/envs/pax_v6.1.0/lib/python3.4/site-packages/cax-4.10.1-py3.4.egg/cax/tasks/process.py", line 98, in _process
    parallel.multiprocess_locally(n_cpus=ncpus, **pax_kwargs)
  File "/project/lgrandi/anaconda3/envs/pax_v6.1.0/lib/python3.4/site-packages/pax-6.1.0-py3.4.egg/pax/parallel.py", line 205, in multiprocess_locally
    traceback)
pax.exceptions.EventBlockHeapSizeExceededException: Pax multiprocessing crashed due to exception in one of the workers. Dumping traceback:
Traceback (most recent call last):
  File "/project/lgrandi/anaconda3/envs/pax_v6.1.0/lib/python3.4/site-packages/pax-6.1.0-py3.4.egg/pax/parallel.py", line 356, in safe_processor
    Processor(**kwargs)
  File "/project/lgrandi/anaconda3/envs/pax_v6.1.0/lib/python3.4/site-packages/pax-6.1.0-py3.4.egg/pax/core.py", line 186, in __init__
    self.run()
  File "/project/lgrandi/anaconda3/envs/pax_v6.1.0/lib/python3.4/site-packages/pax-6.1.0-py3.4.egg/pax/core.py", line 310, in run
    total=self.number_of_events)):
  File "/project/lgrandi/anaconda3/envs/pax_v6.1.0/lib/python3.4/site-packages/pax-6.1.0-py3.4.egg/pax/plugins/io/Queues.py", line 97, in get_events
    self.max_blocks_on_heap, block_id + 1))
pax.exceptions.EventBlockHeapSizeExceededException: We have received over 250 blocks without receiving the next block id (1739) in order. Likely one of the block producers has died without telling anyone.

cax_v4.10.1 - 2016-11-24 04:15:23,149 [ERROR] Pax multiprocessing crashed due to exception in one of the workers. Dumping traceback:
Traceback (most recent call last):
  File "/project/lgrandi/anaconda3/envs/pax_v6.1.0/lib/python3.4/site-packages/pax-6.1.0-py3.4.egg/pax/parallel.py", line 356, in safe_processor
    Processor(**kwargs)
  File "/project/lgrandi/anaconda3/envs/pax_v6.1.0/lib/python3.4/site-packages/pax-6.1.0-py3.4.egg/pax/core.py", line 186, in __init__
    self.run()
  File "/project/lgrandi/anaconda3/envs/pax_v6.1.0/lib/python3.4/site-packages/pax-6.1.0-py3.4.egg/pax/core.py", line 310, in run
    total=self.number_of_events)):
  File "/project/lgrandi/anaconda3/envs/pax_v6.1.0/lib/python3.4/site-packages/pax-6.1.0-py3.4.egg/pax/plugins/io/Queues.py", line 97, in get_events
    self.max_blocks_on_heap, block_id + 1))
pax.exceptions.EventBlockHeapSizeExceededException: We have received over 250 blocks without receiving the next block id (1739) in order. Likely one of the block producers has died without telling anyone`
```

_Copied from original issue: XENON1T/cax#60_

**Comment by pdeperio** [comment]
Full log here: 
```
/project/lgrandi/xenon1t/cax/4696_v6.1.0/cax.log
```
in case you want to test only events around crash.

**Comment by pietrodigangi** [comment]
List of **TPC runs showing transfer (or processing) errors** since 23/11/2016:
4916
4913
4895
4893
4858
4832
4830
4805

looking at log seems that usually at some point one of the logs has errors like
```
pax.exceptions.EventBlockHeapSizeExceededException: We have received over 250 blocks without receiving the next block id (8524) in order. Likely one of the block producers has die
d without telling anyone.
```

and then the following logs has
```
FileNotFoundError: [Errno 2] No such file or directory: '/project/lgrandi/anaconda3/envs/pax_v6.1.0/bin/conda'
```

**Comment by pietrodigangi** [comment]
List of **MV** runs showing transfer (or processing) errors since 23/11/2016:
161124_0030
161124_0630
161124_1202
161124_1446
161124_1510
161124_1517
161124_1547
161124_1619
161124_1707
161124_2330
161125_0530
161125_0955
161125_1058
161125_1202
161125_1356
161125_1620
161125_1652
161125_1710
161125_1728
161125_1915
161126_1407
161126_1413
161127_0213
161127_1415
161128_0215

**Comment by pietrodigangi** [comment]
I tried to manually process the MV datasets
161124_0030 and 161128_0215.

I used the command on midway:
`paxer --input /project/lgrandi/xenon1t/raw/${run_name} --input_type zbson --config XENON1T_MV --config_str [Queues]max_blocks_on_heap=1000 --output myprocess_${run_name}`

and it seems that all went fine.

**Comment by pietrodigangi** [comment]
I also processed manually one of the **TPC** runs (4916) with errors. I used the command line

`paxer --input /project/lgrandi/xenon1t/raw/${run_name} --config XENON1T --config_str [Queues]max_blocks_on_heap=1000 --output myprocess_${run_name}`

The processing in batch job on midway successfully ended after ~13h.
The dataset seems to be **processed correctly**.

**Comment by pdeperio** [comment]
Thanks @pietrodigangi !

@JelleAalbers Could you please check and approve https://github.com/XENON1T/cax/pull/61 ?

**Comment by pdeperio** [comment]
Unfortunately, Run 4696 is still failing with with https://github.com/XENON1T/cax/pull/61:
```
pax.exceptions.QueueTimeoutException: Waited for more than 300 seconds to receive events; lost confidence they will ever come.
```
full log: ```/project/lgrandi/xenon1t/cax/4696_v6.1.0/4696_v6.1.0_22085853.log```. 

Problematic events seems to be around 104, also shown in previous logs (except one time it managed then failed at a later event ~2050-2056).

Please assign high priority for @Ludwig1987's urgent PMT analysis.

**Comment by JelleAalbers** [comment]
Did you try single-core processing (of the affected events)? If that fails to there is a separate issue.

**Comment by JelleAalbers** [comment]
From the log it seems the root class compilation starts very late into the event (almost ten minutes). This explains why the output queue is not moving, and hence why we get the error. 
    * Is it possible not enough cpus were allocated? To run fully smoothly, pax needs 2 + the number of cpus you specify in the cpus argument to work nicely, since cpus only controls the number of workers. Probably 1 + cpus will suffice, input is not very CPU intensive (nor is output -- when the compilation has succeeded). 
    * It is possible to precompile the root class before we start cax jobs?

**Comment by Ludwig1987** [comment]
I am note sure what I am supposed to write here but when I used this command: 

paxer --input /project/lgrandi/xenon1t/raw/161118_1234 --config XENON1T --config_str [Queues]max_blocks_on_heap=1000 --output rootfile/161118_1234

run 4696 was processed on midway without any problem in my home folder.

**Comment by JelleAalbers** [comment]
Thanks Ludwig! So single-core processing works fine (that's what you get without the --cpus argument), in that case even the Queues argument override is unnecessary (queues aren't used for single core processing).

So the question remains why the compilation of the ROOT classes is taking so long on the batch queue that the multiprocessing timeouts are reached -- if indeed this is the cause of the problem.

**Comment by pdeperio** [comment]
@JelleAalbers How can you tell it's a compilation issue from the logs? Looks like it manages to get through some events...

@Ludwig1987 I've started a single core job that should get it produced in the standard location and registered on RunsDB. Could you keep an eye on the job (22088206)? The commands I used for expert reference: 
```
cax-rm --location  /project/lgrandi/xenon1t/processed/pax_v6.1.0/161118_1234.root
qsub /project/lgrandi/xenon1t/cax/4696_v6.1.0/submit.sh
```

**Comment by tunnell** [comment]
Should we be running pax/cax in a temporary local directory?  It could be that it's trying to write to some network mounted storage that is super slow.

**Comment by pdeperio** [comment]
Yes, there should be local scratch space. Now jobs are trying to read/write/compile in /project and /home/tunnell

Is there a way to specify where to compile the libraries?

**Comment by tunnell** [comment]
GRID_MODE or something in rootpy?  can somebody create a rootpy issue and ask?

**Comment by tunnell** [comment]
I think this was due to overly slow disk access at midway.  Processing works again.  After it finishes processing, will close this.

**Comment by tunnell** [comment]
Closing issue since not real issue.  If disks freeze, can't write output, so pax crashes.

**Comment by pdeperio** [comment]
Re-opening since not where else to keep track of this, but still getting this error a significant fraction of the time, preventing the processing of important runs (for monitoring Kr), e.g. in ```/project/lgrandi/xenon1t/cax/5024_v6.1.0/cax.log```:
```
pax.exceptions.QueueTimeoutException: Blocked from pushing to the queue for more than 300 seconds; lost confidence we will ever be able to.
```

**Comment by pdeperio** [comment]
@JelleAalbers also, is there some problem with https://github.com/XENON1T/cax/blob/master/cax/tasks/process.py#L100

pax seems to not be picking up the ```600 seconds```?

**Comment by tunnell** [comment]
But are you sure it's a pax bug?  When pax dies, can you add a command to see if Midway disks are still active?

**Comment by pdeperio** [comment]
Just died again now and disk seems fine, e.g. ```/project/lgrandi/xenon1t/cax/5025_v6.1.0/5025_v6.1.0_22207657.log```

**Comment by JelleAalbers** [comment]
Not picking up the timeouts is because the overrides here https://github.com/XENON1T/pax/blob/master/pax/parallel.py#L138 have higher priority; I hope 966a9edab314110ed16aa899a963dbbf43cbe292 fixes this.

However, looking at your log it is still the same problem: the ROOT compilation is taking a very long time. You can see this because ROOT statements (complaining about liblzma.so, generating dictionary,  etc) which normally occur shortly after each other during compilation are interspersed throughout the log.

Interestingly there is also the following message:
```
ROOT.TGClient.TGClient: WARNING  can't open display "localhost:21.0", switching to batch mode...
```
which seems to indicate it is trying to open a display. I've tried to fix this in 9ecd5c0b94a295893f0bc4eeb0be31039019cd9c. This could be contributing to the delay, but it doesn't seem to be the main cause (looking at the time between the other ROOT statements everything in the compilation just seems to be very slow). 

As I mentioned above, I would suggest two things:
  * Allocate one more CPU in the batch task, but don't add it to the cpus argument of pax. That argument  controls the number of worker processes, but during compilation there is additional CPU-intensive work needed in the output process. It could be this simply doesn't get done because the worker processes are using all cpu.
  * Precompile the pax class library before you start processing. This is a bit unpleasant but may be a workaround. For example, you could process 1 event in single-core mode before you launch the main processing job.

**Comment by pdeperio** [comment]
Thanks Jelle, I'll try these.

But regarding the second suggestion: do you know why the pax_event_class ROOT libraries are compiled with a different hash number each time? Would it be possible to have ```paxer``` be able to load a single copy? I guess if we run all the jobs in the same directory, they should pick up the existing library files. But then I remember we had issues with the jobs getting into race conditions, so we put each job into a separate directory, but this should not be necessary anymore?

**Comment by JelleAalbers** [comment]
Yes, you don't need to put each job in a separate directory anymore. In #439 (see under 'ROOT compilation lockfiles') pax got a way of resolving completely independent complication requests in the same directory: that already happens during multiprocessing. 

I'm not sure if the hashes for the pax_event_class_xxx.cpp libraries are different each time, but for sure the ROOT dictionaries will have different hashes and be recompiled each time if you move to a new folder whether or not this is the case. This is because rootpy (which compiles something for the dictionary) also computes a hash, for which we feed in the full absolute path of the pax_event_class_xxx. See #323 for more information.

We could try to pass a relative pathname here https://github.com/XENON1T/pax/blob/master/pax/plugins/io/ROOTClass.py#L433 instead. That might run into different problems, however; if two processes from the same user start the same dictionary compilation request, there is no longer a negotiation via lockfiles, so one may end up loading the partial dictionary created by the other and crash.

It could be that I'm confusing different compilation steps and hashing mechanisms. I definitely don't remember all the detailed considerations we went through at the time (I remember something with trying to create a temporary directory each compilation, but that didn't work for some reason). ROOT support is one of the most complicated (and tedious) parts of pax.

**Comment by feigaodm** [comment]
Fixed in https://github.com/XENON1T/cax/issues/70
